src/main/java/de/tum/in/www1/artemis/repository/CourseRepository.java

```diff
@@ -116,9 +116,9 @@ SELECT CASE WHEN (count(c) > 0) THEN true ELSE false END
             SELECT DISTINCT c
             FROM Course c
                 LEFT JOIN FETCH c.tutorialGroups tutorialGroups
-                LEFT JOIN FETCH tutorialGroups.teachingAssistant tutor
-                LEFT JOIN FETCH tutorialGroups.registrations registrations
-                LEFT JOIN FETCH registrations.student student
+                LEFT JOIN tutorialGroups.teachingAssistant tutor
+                LEFT JOIN tutorialGroups.registrations registrations
+                LEFT JOIN registrations.student student
```

Comment:
I don't think this will work, without the fetch these values are only used for evaluating the where clause, but won't be present in the result. To enhance the performance we should check the method calls and make sure that tutorial groups are only loaded when necessary.

Reply:
I don't think we will need the TA, registrations, or student in the result, right? We will have the tutorial groups we're looking for, tho - it still has the fetch clause.

Reply:
Ah true I missed that.

Looking at the usages we might improve this even more. This method is only used at one place:
```java
public List<TutorialGroup> findAllForNotifications(User user) {
        return courseRepository.findAllActiveWithTutorialGroupsWhereUserIsRegisteredOrTutor(ZonedDateTime.now(), user.getId()).stream()
                .flatMap(course -> course.getTutorialGroups().stream()).toList();
    }
```

which is again only used here:
```java
        var tutorialGroupIds = tutorialGroupService.findAllForNotifications(currentUser).stream().map(DomainObject::getId).collect(Collectors.toSet());
``` 

So in the end we load the whole course but only need the ids of the tutorial groups:

```java
 @Query("""
            SELECT DISTINCT tutorialGroups.id
            FROM Course c
                LEFT JOIN FETCH c.tutorialGroups tutorialGroups
                LEFT JOIN tutorialGroups.teachingAssistant tutor
                LEFT JOIN tutorialGroups.registrations registrations
                LEFT JOIN registrations.student student
            WHERE (c.startDate <= :now OR c.startDate IS NULL)
                AND (c.endDate >= :now OR c.endDate IS NULL)
                AND (student.id = :userId OR tutor.id = :userId)
            """)
    Set<Long> findAllActiveTutorialGroupIdsWhereUserIsRegisteredOrTutor(@Param("now") ZonedDateTime now, @Param("userId") Long userId);
```
(not sure if this will work, I did not test it)

Reply:
=> Followup PR :)